gstd/src/common/handlers.rs

```diff
@@ -25,50 +25,208 @@
 //! debug and non-debug mode, for programs built in `wasm32` architecture.
 //! For `debug` mode it provides more extensive logging.
 
+#[cfg(target_arch = "wasm32")]
 #[cfg(feature = "oom-handler")]
 #[alloc_error_handler]
 pub fn oom(_: core::alloc::Layout) -> ! {
     crate::ext::oom_panic()
 }
-
+/// We currently support 3 panic handler profiles:
+/// - `panic-handler`: it displays `panicked with '<unknown>'`
+/// - `panic-message`: it displays `panicked with '{message}'`
+/// - `panic-location`: it displays `panicked with '{message}' at '{location}'`
+///
+/// How we get the panic message in different versions of Rust:
+/// - In nightly Rust, we use `#![feature(panic_info_message)]` and the
+///   [`write!`] macro.
+/// - In stable Rust, we need to modify the default panic handler message
+///   format.
+///
+/// Default panic handler message format (according to <https://github.com/rust-lang/rust/pull/112849>):
+/// - Rust  <1.73: `panicked at '{message}', {location}`
+/// - Rust >=1.73: `panicked at {location}:\n{message}`
+///
+/// We parse the output of `impl Display for PanicInfo<'_>` and
+/// then convert it to custom format:
+/// `panicked with '{message}'[ at '{location}']`.
+#[cfg(target_arch = "wasm32")]
 #[cfg(feature = "panic-handler")]
 mod panic_handler {
     use crate::ext;
     use core::panic::PanicInfo;
 
-    #[cfg(not(feature = "debug"))]
+    mod constants {
+        /// This prefix is used before the panic message.
+        pub const PANIC_PREFIX: &str = "panicked with ";
+        /// This panic message is used in the minimal panic handler and when
+        /// internal errors occur.
+        #[cfg(any(feature = "panic-info-message", not(feature = "panic-message")))]
+        pub const UNKNOWN_REASON: &str = "<unknown>";
+
+        /// This prefix is used by `impl Display for PanicInfo<'_>`.
+        #[cfg(all(not(feature = "panic-info-message"), feature = "panic-message"))]
+        pub const PANICKED_AT: &str = "panicked at ";
+
+        /// Max amount of bytes allowed to be thrown as string explanation
+        /// of the error.
+        #[cfg(feature = "panic-message")]
+        pub const TRIMMED_MAX_LEN: usize = 1024; //TODO: do not duplicate `gear_core::str::TRIMMED_MAX_LEN`
+        /// Size of array string that will be allocated on the stack.
+        #[cfg(feature = "panic-message")]
+        pub const ARRAY_STRING_MAX_LEN: usize = PANIC_PREFIX.len() + TRIMMED_MAX_LEN;
+    }
+
+    use constants::*;
+
+    /// Minimal panic handler.
+    #[cfg(not(feature = "panic-message"))]
     #[panic_handler]
     pub fn panic(_: &PanicInfo) -> ! {
-        ext::panic("no info")
+        const MESSAGE: &str = const_format::formatcp!("{PANIC_PREFIX}'{UNKNOWN_REASON}'");
```

Comment:
The whole dependency can be replaced with `concat!()`
```suggestion
        const MESSAGE: &str = concat!(PANIC_PREFIX, "'", UNKNOWN_REASON, "'");
```

Reply:
https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=812db8d01033f6f02bef668f38193c46
it doesn't compile